### PR TITLE
Vm delete

### PR DESCRIPTION
**Pull request by gu-bin**
Moved post checking active transactions for delete vm to a private method

### CODE REVIEW

File: softlayer/vm/softlayer_vm.go [hunk 1]
```diff
@@ -279,3 +290,61 @@ func (vm SoftLayerVM) getRootPassword(virtualGuest datatypes.SoftLayer_Virtual_G
...
+			vm.logger.Info(deleteVMLogTag, "VM doesn't exist. Delete done", nil)
+			break
+		}
+
+		activeTransaction, err := virtualGuestService.GetActiveTransaction(virtualGuestId)
+		if err != nil {
+			return bosherr.WrapError(err, "Getting active transactions from SoftLayer client")
+		}
+
+		averageTransactionDuration, _ := strconv.ParseFloat(activeTransaction.TransactionStatus.AverageDuration, 32)
```
Comment (maximilien) [inline softlayer/vm/softlayer_vm.go]: Why are you ignoring the `error` of `strconv.ParseFloat`. What if SL sends garbage and the ParseFloat fails.. You need to check `if err != nil` and either recover (e.g., retry or something like that) or return and fail.

**Comment by maximilien** [comment]
Hi, @gu-bin,

OK I have one comment. Please address.

Also, can we have a test that times out. You could set the timeout value low and do a test that timeout by sleeping in the Fake. The purpose of such a test would be to indirectly test the `vm.postCheckActiveTransactionsForDeleteVM` since as you make it private there are no real test for it. Right?

Please address above and I will merge. I am not closing PR so you can make update. Just make sure to add a comment so I know to review.

Thanks,

@maximilien

**Comment by gu-bin** [comment]
@maximilien 
To test vm.postCheckActiveTransactionsForDeleteVM, this is my understanding. Is it correct?
1. Copy method postCheckActiveTransactionsForDeleteVM of softlayer_vm.go to fake_vm.go as fakePostCheckActiveTransactionsForDeleteVM and add sleep sentences to it.
2. In FakeVM.Delete(), only simply invoke fakePostCheckActiveTransactionsForDeleteVM with low timeout value passing in.
3. In softlayer_vm_test.go, add a test case to test the timeout by invoking FakeVM. fakePostCheckActiveTransactionsForDeleteVM.

**Comment by maximilien** [comment]
@gu-bin, there are many ways to do this. However, we need to be careful not to fake too many things since the test will be useless.

I would suggest to start with:

1. add a means to change the `timeout` upper bound. This is so that it can be set to a small value, say 5 seconds.
2. in your `VM.Delete()` you should then just use the fake client which will load test fixtures normally. 
3. add a way in the fake client to specify how long to sleep for a `DoXyz` call
4. setup the new test so that your timeout value is less than the total sleep for the fake client

That way nothing else has to change. OK?

**Comment by gu-bin** [comment]
@maximilien I've update the code according to your comments. Please review again. Thanks.

File: softlayer/vm/softlayer_vm.go [hunk 2]
```diff
@@ -32,6 +34,8 @@ type SoftLayerVM struct {
 	sshClient util.SshClient
 
 	logger boshlog.Logger
+
+	timeOutTestTag bool
```
Comment (maximilien) [inline softlayer/vm/softlayer_vm.go]: No. So This is not the way to do this. Instead, inject the actual timeout as a parameter when creating the VM. That way you can provide the right value when creating the VM in test with the fake client.

You can still keep the default timeout as public constant in the SoftLayerVM and use that as needed when creating a new VM in the create VM implementation.

**Comment by maximilien** [comment]
We are getting closer. See last comment. Thanks.

**Comment by maximilien** [comment]
So this is close, correct idea but I feel that the `timeout` not the bool should be injected. That way the code is simple, you are checking if this is running as test or not. All you do is set the correct timeout value and run. In the test you set the timeout value small and in normal code you use the defaul value.

If you need me to comment in speudocode to do this then let me know.

**Comment by gu-bin** [comment]
@maximilien Did you mean adding a ```timeout``` parameter to the existing ```NewSoftLayerVM``` method? I think in this way all the calling to this method need to be modified. Or I need to create a new method to create VM and add the timeout parameter? If this is not correct, please provide the speudocode as I'm pretty new to golang. Thanks.

**Comment by maximilien** [comment]
Yes, correct. Add parameter to the factory method (the `New...` method) and modify all callers. That's what I mean by inject the timeout parameter. Don't just call it timeout but be specific about what its use is, so maybe: `timeoutCheckingForActiveTransactions` or similar. 

Also, make the default value a constant that the caller of the newly modified method can use to inject, like I did for: `bslcommon.TIMEOUT` But of course a better name, e.g., `TIMEOUT_TRANSACTIONS_DELETE_VM` Or similar.

This idea of injection is very common. Please google it. Typically things that need to vary for testing purposes are injected into code to make that testing easier. It is not a Golang concept but is a generic concept. Especially useful for TDD.

**Comment by gu-bin** [comment]
@maximilien please review again. Thanks.

File: softlayer/vm/softlayer_creator.go [hunk 3]
```diff
@@ -95,7 +95,7 @@ func (c SoftLayerCreator) Create(agentID string, stemcell bslcstem.Stemcell, clo
 
 	agentEnvService := c.agentEnvServiceFactory.New(virtualGuest.Id)
 
-	vm := NewSoftLayerVM(virtualGuest.Id, c.softLayerClient, util.GetSshClient(), agentEnvService, c.logger)
+	vm := NewSoftLayerVM(virtualGuest.Id, c.softLayerClient, util.GetSshClient(), agentEnvService, c.logger, TIMEOUT_TRANSACTIONS_DELETE_VM)
```
Comment (maximilien) [inline softlayer/vm/softlayer_creator.go]: Nice, that's exactly what I wanted. Thanks.

File: softlayer/vm/softlayer_vm.go [hunk 4]
```diff
@@ -279,3 +297,66 @@ func (vm SoftLayerVM) getRootPassword(virtualGuest datatypes.SoftLayer_Virtual_G
...
+			break
+		}
+
+		vm.logger.Info(deleteVMLogTag, "This is a short transaction, waiting for all active transactions to complete", nil)
+		totalTime += pollingInterval
+		time.Sleep(pollingInterval)
+	}
+
+	if totalTime >= timeout {
+		err := errors.New(fmt.Sprintf("After deleting a vm, waiting for active transactions to complete TIME OUT!"))
```
Comment (maximilien) [inline softlayer/vm/softlayer_vm.go]: This could be one line since the next is simply `return err`

**Comment by maximilien** [comment]
Nice @gu-bin, well done. I had one minor comment but merging this and you can address in future.

Let's now monitor @posijon's pipeline to make sure there are no issues with integration test.

Thanks.

**Pull request merged**